common/replay/replay_guardian.go

```diff
@@ -0,0 +1,169 @@
+package replay
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/emirpasic/gods/queues/priorityqueue"
+)
+
+// ReplayGuardian ensures that the same request is not processed more than once. It can be used to do things such
+// as protecting against replay attacks or accidental duplicate requests.
+type ReplayGuardian interface {
+
+	// VerifyRequest verifies that a request with the given hash and timestamp is not a replay
+	// of a previous request. If it cannot be determined if a request is a replay or not,
+	// then the request is rejected. Only if it can be guaranteed that the request is not a replay
+	// will this method return nil.
+	//
+	// In order to be a verified unique request, the following conditions must be met:
```

Comment:
It seems this description after the first paragraph is implementation specific - given that there is an interface defined, this doc might be better placed alongside the implementation the contains this logic

Reply:
Makes sense, the second paragraph has been removed from the interface.